fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BatchServiceImpl.java

```diff
@@ -0,0 +1,135 @@
+ * limitations under the License.
+ */
+/**
+ *
+ */
+
+package org.fcrepo.kernel.modeshape.services;
+
+import static java.time.Instant.now;
+import static java.util.stream.Collectors.toSet;
+import static org.fcrepo.kernel.modeshape.FedoraSessionImpl.operationTimeout;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.fcrepo.kernel.api.FedoraSession;
+import org.fcrepo.kernel.api.exception.RepositoryRuntimeException;
+import org.fcrepo.kernel.api.exception.SessionMissingException;
+import org.fcrepo.kernel.api.services.BatchService;
+
+import org.slf4j.Logger;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+/**
+ * This is part of the strawman implementation for Fedora batch operations This
+ * service implements a simple {@link FedoraSession} service which is able to
+ * create/commit/rollback {@link FedoraSession} objects. A {@link Scheduled}
+ * annotation is used for removing timed out operations
+ *
+ * @author frank asseg
+ * @author ajs6f
+ * @author acoburn
+ */
+@Component
+public class BatchServiceImpl extends AbstractService implements BatchService {
+
+    private static final Logger LOGGER = getLogger(BatchServiceImpl.class);
+
+    /**
+     * TODO since sessions have to be available on all nodes, they have to
+     * be either persisted or written to a distributed map or sth, not just this
+     * plain hashmap that follows
+     */
+    private static Map<String, FedoraSession> transactions = new ConcurrentHashMap<>();
+
+    public static final long REAP_INTERVAL = 1000;
+
+    /**
+     * Every REAP_INTERVAL milliseconds, check for expired sessions. If the
+     * tx is expired, roll it back and remove it from the registry.
+     */
+    @Override
+    @Scheduled(fixedRate = REAP_INTERVAL)
+    public void removeExpired() {
+        final Set<String> reapable = transactions.entrySet().stream()
+                .filter(e -> e.getValue().getExpires().isBefore(now())).map(Map.Entry::getKey).collect(toSet());
+        reapable.forEach(key -> {
+            final FedoraSession s = transactions.get(key);
+            if (s != null) {
+                try {
+                    s.expire();
+                } catch (final RepositoryRuntimeException e) {
+                    LOGGER.error("Got exception rolling back expired session {}: {}", s, e.getMessage());
+                }
+            }
+            transactions.remove(key);
+        });
+    }
+
+    @Override
+    public void begin(final FedoraSession session, final String username) {
+        transactions.put(getTxKey(session.getId(), username), session);
+    }
+
+    @Override
+    public FedoraSession getSession(final String sessionId, final String username) {
+        final FedoraSession tx = transactions.computeIfAbsent(getTxKey(sessionId, username), s -> {
```

Comment:
The Javadocs for `Map::computeIfAbsent` say `The default implementation makes no guarantees about synchronization or atomicity properties of this method.` so you are relying on `ConcurrentHashMap`s impl-specific behavior here.... but do we have any understanding anyway of the atomicity of `*Service`? I don't think this is coherent.